### ISSUE TITLE
Lines beginning with hash symbol in literal code blocks treated as headers

### ISSUE THREAD

**Issue by duncanburke** [jrblevin/markdown-mode#33]
For the purposes of commands such as `outline-next-visible-heading` and `outline-forward-same-level`, it is not expected that lines beginning with the `#` symbol in literal code blocks be treated as headers.

For example:

    ### Third level ###

    * A list item
    * and another
      with some overhang
       + and a sublist
         with overhang
    - And a third with a different marker

    ### Third level number two ###

    ~~~ {.bash}
    #!/bin/bash

    # In a code block, this shouldn't be a header
    ~~~

    ### Level two again

    followed by some body text

**Comment by jrblevin** [comment]
I think the new `markdown-collect-fm-code-blocks` function introduced in #40 could be used for font lock to fix this issue also. By identifying fenced code blocks first, then identifying headings that are not already marked as code blocks. I'll try to take a look soon, but if anyone can get to it sooner please submit a pull request.

**Comment by syohex** [comment]
In `outline-next-visible-heading`, outline-mode uses only regular expression(`outline-regexp`) for searching header. We can control its behaviour only by `outline-regexp`. So we need to write wrapper command for avoiding this issue as below.

```lisp
(defun markdown-move-visible-header-common (move-fn arg)
  (let ((blocks (markdown-collect-gfm-code-blocks)))
    (funcall move-fn arg)
    (while (markdown-gfm-code-block-p blocks (point))
      (funcall move-fn arg))))

(defun markdown-next-visible-heading (arg)
  (interactive "p")
  (markdown-move-visible-header-common #'outline-next-visible-heading arg))

(defun markdown-next-visible-heading (arg)
  (interactive "p")
  (markdown-move-visible-header-common #'outline-previous-visible-heading arg))
```

I don't understand `outline-forward-same-level` behavior yet :-(

**Comment by syohex** [comment]
I sent PR #41. Please check it.

**Comment by jrblevin** [comment]
Thanks for this report, @duncanburke, and thanks @syohex for the patch. As I mentioned in #41 I have been working on syntax table support for code blocks. In addition to helping with font lock, this is also related to the heading movement commands. I used #41 as a basis for the changes. So, this should be fixed in 5daa9fc.
